base/multidimensional.jl

```diff
@@ -364,116 +383,57 @@ end
 # In such cases we have to collapse the 2d space spanned by the ranges.
 #
 # API:
-#    merge_indexes(V, indexes::NTuple, dims::Dims, linindex)
-# where dims encodes the trailing sizes of the parent array,
+#    merge_indexes(V, indexes::NTuple, index)
 # indexes encodes the view's trailing indexes into the parent array,
-# and linindex encodes the subset of these elements that we'll select.
-#
-# The generic algorithm makes use of div to convert elements
-# of linindex into a cartesian index into indexes, looks up
-# the corresponding cartesian index into the parent, and then uses
-# dims to convert back to a linear index into the parent array.
+# and index encodes the subset of these elements that we'll select.
 #
-# However, a common case is linindex::Range.
-# Since div is slow and in(j::Int, linindex::Range) is fast,
-# it can be much faster to generate all possibilities and
-# then test whether the corresponding linear index is in linindex.
-# One exception occurs when only a small subset of the total
-# is desired, in which case we fall back to the div-based algorithm.
-#@generated function merge_indexes{T<:Integer}(V, parentindexes::NTuple, parentdims::Dims, linindex::Union{Colon,Range{T}}, lindim)
-@generated function merge_indexes_in{TT}(V, parentindexes::TT, parentdims::Dims, linindex, lindim)
-    N = length(parentindexes.parameters)   # number of parent axes we're merging
-    N > 0 || throw(ArgumentError("cannot merge empty indexes"))
-    lengthexpr = linindex == Colon ? (:(prod(size(V)[lindim:end]))) : (:(length(linindex)))
-    L = symbol(string("Istride_", N+1))  # length of V's trailing dimensions
+# It returns a CartesianIndex or array of CartesianIndexes.
```

Comment:
Just in case the memory issue is not apparent:
```jl
A = rand(10,10,10)
B = sub(A, 2:999)
```
`A` consumes 8000 bytes. In `master`, `B` is about the same size (because `sizeof(Int) == sizeof(Float64)`. In this PR, `B` consumes approximately 24000 bytes, so the view is considerably larger than the original.